### ISSUE TITLE
isSimple fails in some cases

### ISSUE THREAD

**Issue by reinvantveer** [bjornharrtell/jsts#205]
Maybe I'm missing something, but it seems that whatever you throw at it, isSimple(); always returns true for Simple Feature Spec validation. From [Wikipedia](https://en.wikipedia.org/wiki/Simple_Features):
"In general, a 2D geometry is simple if it contains no self-intersection." 

Yet, with some testing, self-intersecting polygons can be shown to be false for the validation isValid(); but at the same time true for isSimple();

See [gist](https://gist.github.com/reinvantveer/8c3dd817c70ff0304482) for the test script. It looks straightforward enough to me but maybe I'm missing something.

Thanks in advance!

**Comment by dr-jts** [comment]
It looks like the JSTS code reflects an older version of the JTS IsSimple.  In Sept 2012 (https://sourceforge.net/p/jts-topo-suite/code/662/) the JTS IsSimple was changed to check for self-intersecting polygon rings.

**Comment by bjornharrtell** [comment]
I've updated the gist to work against the new 1.x branch which should reflect an up to date JTS [here](https://gist.github.com/bjornharrtell/1b60e929eb1c0fa31fea). It seems the issues remain. The output is:

```
Processing feature: 0
Self-intersection found at GeoJSON point: {"type":"Point","coordinates":[6.587385486997261,52.58402645716811]}
Is a Simple Feature: true
Is a valid feature: false

Processing feature: 1
No self-intersection found
Is a Simple Feature: true
Is a valid feature: true

Processing feature: 2
Self-intersection found at GeoJSON point: {"type":"Point","coordinates":[5.864005862171732,53.093803816940195]}
Is a Simple Feature: true
Is a valid feature: false

Processing feature: 3
No self-intersection found
Is a Simple Feature: true
Is a valid feature: true
```

**Comment by reinvantveer** [comment]
That's too bad, as thus can be considered an important function. A lot of spatially enabled data stores, such as elasticsearch, need features to conform to standards much stricter than the geojson spec itself.

**Comment by bjornharrtell** [comment]
Yes, I'd like to have it fixed before releasing 1.0.0 as stable. It would help to come up with a simple as possible test case.

**Comment by bjornharrtell** [comment]
To be noted, the current master does pass tests that involve self-intersections, for example the test case TestSimple - LR - ring with self-intersection is passing with an expected `false` from `isSimple`. The input geometry in that case is `LINEARRING (100 300, 200 300, 100 200, 200 200, 100 300)`.

**Comment by reinvantveer** [comment]
Is the gist I sent earlier of enough use? Or do you need more?

Op vr 5 feb. 2016 20:56 schreef Björn Harrtell <notifications@github.com>:

**Comment by bjornharrtell** [comment]
Yes there are two problems. Your gist is too complex both code wise and geometries given. Example what I mean with simple as possible test cases:

```js
var g1 = reader.read('LINESTRING(0 0, 10 0)')
console.log(g1.isSimple())
var g2 = reader.read('LINESTRING(0 0, 10 0, 2.5 5, 7.5 5, 0 0)')
console.log(g2.isSimple())
```

The above simple cases gives true for g1 and false for g2 (which has an obvious self-intersection). I need to find a similar simple geometry example that fails to get to the bottom of what is causing it. Also, any case should also be run in JTS to make sure where the issue is.

**Comment by bjornharrtell** [comment]
Seems I might have found a good example.

```js
var g3 = reader.read('POLYGON((0 0, 10 0, 2.5 5, 7.5 5, 0 0))')
console.log(g3.isSimple())
```

By just making the linestring a polygon it seems JSTS fails to detect it as non-simple. In JTS is it reported as non-simple.

**Comment by bjornharrtell** [comment]
The issue is at https://github.com/bjornharrtell/jsts/blob/master/src/org/locationtech/jts/operation/IsSimpleOp.js#L130 where the instanceof check fails to detect the Polygon as implementing the Polygonal interface. Need consideration in the AST translation.

**Comment by bjornharrtell** [comment]
Should be fixed as of 9c7c28b7b087fac5f9df52107737b564ccbd067c.

**Comment by reinvantveer** [comment]
Nice

Op za 6 feb. 2016 12:39 schreef Björn Harrtell <notifications@github.com>:
